hwtracer/src/decode/ykpt/mod.rs

```diff
@@ -19,29 +38,232 @@ impl TraceDecoder for YkPTTraceDecoder {
+                    let target_off = if self.tnts.pop_front().unwrap() {
+                        *taken_target
+                    } else {
+                        if let Some(ntt) = not_taken_target {
+                            *ntt
+                        } else {
+                            // Divergent control flow.
+                            panic!();
+                        }
+                    };
+                    self.cur_loc = ObjLoc::MainObj(target_off);
+                    return Ok(Some(self.lookup_block_from_main_bin_offset(target_off)));
+                }
+                SuccessorKind::Dynamic => {
+                    // We can only know the successor via a TIP update in a packet. Fall through to
+                    // `self.seek_tip()`.
+                    self.cur_loc = ObjLoc::OtherObjOrUnknown;
+                }
+            }
+        }
+
+        // If we get here then we can't statically figure out where to go next. A packet is going
+        // to have to tell us what to do.
+        self.seek_tip()
+    }
+
+    /// Keep decoding packets until we have some TNT decisions buffered.
+    fn seek_tnt(&mut self) -> Result<bool, HWTracerError> {
+        while self.tnts.is_empty() {
+            let pkt = self.packet()?; // Potentially populates `self.tnts`.
+            if pkt.is_none() {
+                return Ok(false); // Packet stream exhausted.
+            }
+        }
+        Ok(true)
+    }
+
+    /// Keep decoding packets until we receive a TIP update.
+    fn seek_tip(&mut self) -> Result<Option<Block>, HWTracerError> {
+        debug_assert_eq!(self.cur_loc, ObjLoc::OtherObjOrUnknown);
+        loop {
+            if let Some(_) = self.packet()? {
+                if let ObjLoc::MainObj(off) = self.cur_loc {
+                    return Ok(Some(self.lookup_block_from_main_bin_offset(off)));
+                }
+            } else {
+                return Ok(None); // No more packets.
+            }
+        }
+    }
+
+    /// Fetch the next packet and update iterator state.
+    fn packet(&mut self) -> Result<Option<Packet>, HWTracerError> {
+        let ret = if let Some(pkt_or_err) = self.parser.next() {
+            let pkt = pkt_or_err?;
+            if pkt.kind() == PacketKind::TIPPGE {
+                debug_assert!(!self.pge);
+                self.pge = true;
+            }
+            if pkt.kind() == PacketKind::TIPPGD {
+                debug_assert!(self.pge);
+                self.pge = false;
+            }
+            if let Some(vaddr) = pkt.target_ip() {
+                if self.pge {
+                    match code_vaddr_to_off(vaddr) {
+                        Some((obj, off)) if obj == *SELF_BIN_PATH => {
+                            self.cur_loc = ObjLoc::MainObj(off)
+                        }
+                        _ => self.cur_loc = ObjLoc::OtherObjOrUnknown,
+                    }
+                    self.tnts.clear();
```

Comment:
Why do we clear?

Reply:
Because at this point any remaining TNTs would be ones for past control flow decisions that we were unable to make use of due to imperfect knowledge of the control flow.

There are two situations I know of where this happens:

 - execution passed through code that wasn't compiled by ykllvm, so we get TNT packets, but no static successor info to use them against.
 -  At the start of a trace, often a TNT packet comes in prior to the `TIP.PGE` packet indicating the start of tracing. Don't ask we why :P

Reply:
Can we add a comment (or, even better, an assertion) for this?

Reply:
Yes, and I will try :)

Reply:
Fixed in 099ae515c5487c4fda7b7d3f8d5c1ed658fc9b88.

(It's not easy to make an assertion I'm afraid)